### PR TITLE
SPHINX-Carla Boyd -Weixi He 

### PR DESCRIPTION
**Pull request by carlamayor**

### CODE REVIEW

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -158,13 +158,13 @@ def test_moves_movie_from_watchlist_to_empty_watched():
...
-    
-    raise Exception("Test needs to be completed.")
-    # *******************************************************************************************
-    # ****** Add assertions here to test that the correct movie was added to "watched" **********
-    # *******************************************************************************************
 
-@pytest.mark.skip()
+    if updated_data["watched"]:
+        if updated_data["watched"][0]["title"] != MOVIE_TITLE_1:
+            raise ValueError("The wrong movie was moved to the watched list.")
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py]: The assert statement on line 160 `assert len(updated_data["watched"]) == 1` will only pass if there is one element in the `watched` list. This means that the `if` condition  on line 162 `if updated_data["watched"]:` is not necessary because the test would have already failed if `updated_data["watched"]` was empty.

Pytest will raise errors on it own if an `assert` statement fails, so we should not be using the `if`/`else` structure here. We should be writing new assert statements. If we want to confirm that a movie title is a particular value, we could write an assert statement like:
```py
    watched_movie = updated_data["watched"][0]
    assert watched_movie["title”] == MOVIE_TITLE_1
```

Additionally, in this case where the test is checking individual keys of a dictionary like `"title:`, I recommend including assertions for all of the relevant keys.
```py
    watched_movie = updated_data["watched"][0]
    assert watched_movie["title”] == MOVIE_TITLE_1
    assert watched_movie["rating"] == RATING_1
    assert watched_movie["genre"] == GENRE_1
```

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -183,12 +183,14 @@ def test_moves_movie_from_watchlist_to_watched():
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
     
-    raise Exception("Test needs to be completed.")
-    # *******************************************************************************************
-    # ****** Add assertions here to test that the correct movie was added to "watched" **********
-    # *******************************************************************************************
+    if updated_data["watched"][-1]["title"] != movie_to_watch["title"]:
+        raise ValueError(f"Incorrect movie added to watched list: expected '{movie_to_watch[0]['title']}', got '{updated_data["watched"][-1]["title"]}'")
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py]: The README isn't explicit about what order we should add movies to the list `watched`. If we want our tests to be independent from a specific implementation that adds movies to a particular location in the `watched` list, we could use the `in` keyword to check if the variable `movie_to_watch` is in the `watched` list instead of looking at a specific index. By checking for the whole dictionary, there's a side benefit that we don't need to check each key independently:
```python
assert movie_to_watch in updated_data["watched"]

# Another option:
assert HORROR_1 in updated_data["watched"]
```

File: tests/test_wave_03.py [hunk 3]
```diff
@@ -55,12 +55,13 @@ def test_friends_unique_movies_not_duplicated():
...
     assert len(friends_unique_movies) == 3
 
-    raise Exception("Test needs to be completed.")
-    # *************************************************************************************************
-    # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
-    # **************************************************************************************************
+    
+    for movie in friends_unique_movies:
+        if friends_unique_movies.count(movie) > 1:
+            raise ValueError(f"Duplicate movie found: {movie}")
```
Comment (kelsey-steven-ada) [inline tests/test_wave_03.py]: The assert on line 56 confirms that there are 3 elements in `friends_unique_movies`. This loop confirms that there is only one of each movie in the list, so we know there are no duplicates, but this doesn't confirm that the 3 movies are the exact movies we expect them to be. 

How could we remove this loop and replace it with assert statements that confirm that the 3 movies we expect are inside the `friends_unique_movies` list?

File: tests/test_wave_05.py [hunk 4]
```diff
@@ -53,12 +53,13 @@ def test_new_genre_rec_from_empty_friends():
...
 
-    raise Exception("Test needs to be completed.")
-    # *********************************************************************
-    # ****** Complete the Act and Assert Portions of these tests **********
-    # *********************************************************************
+    # Act
+    new_recomemendation = get_rec_from_favorites(sonyas_data)
+
+    #Assert
+    assert new_recomemendation == []
```
Comment (kelsey-steven-ada) [inline tests/test_wave_05.py:60]: Nice assertion to confirm we get an empty list back!

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,122 @@
 # ------------- WAVE 1 --------------------
-
+user_data = {
+    "watched": [],
+    "watchlist": []
+}
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: This data structure should be removed. It's likely to cause confusion for readers because it is declared in a global space as if it should be used by multiple functions, however we already have several functions that declare a parameter variable with the same name. When we have a conflict between the name of a local and a global variable, the local variable takes priority over the global variable within the specific scope where it is defined (each function that declares `user_data` as a parameter).

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,122 @@
...
 def create_movie(title, genre, rating):
-    pass
+    movie = {}
+    if title and genre and rating:
+        movie["title"] = title
+        movie["genre"] = genre
+        movie["rating"] = rating
+        return movie
+    else:
+        return None
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:11]: I like that this approach is explicit about what is returned, and nice use of Truthy/Falsy values to check if the parameters are empty or None!

Right now we allocate space for the variable `movie` before we know if the inputs are valid, and the intended flow of our code is indented. We want our goal, our intended actions to be as prominent and clear as possible. I would tend to reverse the sense of this comparison to check if NOT each parameter exists. Then we could treat the if-statement as a guard clause (returning None), and unindent the main logic. 

```py
if not title or not genre or not rating:
    return None
 
movie = {}   
movie["title"] = title
movie["genre"] = genre
movie["rating"] = rating
return movie
```

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,122 @@
...
+def add_to_watchlist(user_data, movie):
+    user_data["watchlist"].append(movie)
+    return user_data
 
+def watch_movie(user_data, title):
+
+    for movie in user_data["watchlist"]:
+        if movie["title"] == title:
+            user_data["watchlist"].remove(movie)
+            user_data["watched"].append(movie)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: This function works, but it contains an "anti-pattern" that is important to be aware of.  An element of the list that is being iterated over is being removed inside the for loop. This won't cause problems in this particular case, but consider this slightly different implementation:
```
def watch_movie(user_data, title):
    for i in range(len(user_data["watchlist"])):
        item = user_data["watchlist"][i]

        if item["title"] == title:
            user_data["watched"].append(item)
            user_data["watchlist"].remove(item)

    return user_data
```
If we were to test this in a scenario where `user_data["watchlist"]` has 2 elements, and the first element is the movie that we are moving to `user_data["watched"]`, during the second iteration through the loop it will throw an IndexError, because we will try to access the 2nd element in `user_data["watchlist"]`, but there no longer is a second element.  

This is a coding pattern that I recommend avoiding because it has so much potential for unforeseen errors. One fix is to use `break` or return `user_data` immediately after making the swap so that the loop ends as soon as the swap is made.  Another fix is to use the loop to find the item, but not make the swap until after the loop ends.

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,122 @@
...
+    if not user_data["watched"]:
+        return 0.0
+    # caculate total rating
+    total_rating = 0
+    for movie in user_data["watched"]:
+        total_rating += movie["rating"]
+
+    avg_rating = total_rating / len(user_data["watched"])
 
+    return avg_rating
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Nice approach! Another way we could do this is using a list comprehension and the `sum` function:
```python
# The comprehension inside the `sum` call pulls the value of the 
# `rating` key out of each `movie` in the list `user_data['watched']`
total = sum(movie["rating"] for movie in user_data["watched"])
average_rating = total / len(user_data["watched"])
return average_rating
```

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,122 @@
...
+    most_genre = {}
+    # bulid a dict for genre count
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre in most_genre:
+            most_genre[genre] +=1
+        else:
+            most_genre[genre] =1
+    # use max to find the most one
+    return max(most_genre, key=most_genre.get)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:60]: Great use of a frequency map and `max`!

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,122 @@
...
+    friends_movies = set()
 
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    user_unique = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_movies:
+            user_unique.append(movie)
+    return user_unique
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: This function is a little visually crowded, I'd recommend adding a new line or two to break up sections of code that are handling different actions. One possibility could be:
```py
    friends_movies = set()
    for friend in user_data["friends"]:
        for movie in friend["watched"]:
            friends_movies.add(movie["title"])
            
    user_unique = []
    for movie in user_data["watched"]:
        if movie["title"] not in friends_movies:
            user_unique.append(movie)
            
    return user_unique
```

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,122 @@
...
+            most_genre[genre] +=1
+        else:
+            most_genre[genre] =1
+    # use max to find the most one
+    return max(most_genre, key=most_genre.get)
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    friends_movies = set()
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: I suggest renaming this something like `friend_movie_titles` to better signify that we aren't holding full movie information in each element, just a movie title.

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,122 @@
...
 
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    user_unique = []
+    for movie in user_data["watched"]:
+        if movie["title"] not in friends_movies:
+            user_unique.append(movie)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Love the use of a `set` here to handle ensuring a collection of unique movie titles with the benefit of O(1) lookup when you loop over the movies in `user_data["watched"]`.

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,122 @@
...
+    return user_unique
+
+def get_friends_unique_watched(user_data):
+    friends_movies = set()
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
+    
+    user_watched = {movie["title"] for movie in user_data["watched"]}
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py:78]: Nice comprehension to make the user set!

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,122 @@
...
+        if movie["title"] not in friends_movies:
+            user_unique.append(movie)
+    return user_unique
+
+def get_friends_unique_watched(user_data):
+    friends_movies = set()
+    
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friends_movies.add(movie["title"])
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: The set `friends_movies` is created and filled with movie titles, but then it isn't referenced again. Since we never read from the `set` after filling it, we should remove this code

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,122 @@
...
+    return friends_unique
         
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    recommended_movies = []
+    if "subscriptions" in user_data and user_data["subscriptions"]:
+        for movie in  get_friends_unique_watched(user_data):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Extra space after the `in` operator
```suggestion
        for movie in get_friends_unique_watched(user_data):
```

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,122 @@
...
+            if movie["host"] in user_data["subscriptions"]:
+                recommended_movies.append(movie)
+    return recommended_movies
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Nice algorithm, love to see the code reuse with `get_friends_unique_watched`!

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,122 @@
...
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
-
+def get_available_recs(user_data):
+    recommended_movies = []
+    if "subscriptions" in user_data and user_data["subscriptions"]:
+        for movie in  get_friends_unique_watched(user_data):
+            if movie["host"] in user_data["subscriptions"]:
+                recommended_movies.append(movie)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Similar to the feedback in `create_movie` I would strongly suggest changing the validation check into a guard clause so that the main flow of the function is as un-indented as possible.
```py
recommended_movies = []
if "subscriptions" not in user_data or not user_data["subscriptions"]:
    return recommended_movies
    
for movie in get_friends_unique_watched(user_data):
    if movie["host"] in user_data["subscriptions"]:
        recommended_movies.append(movie)
                
return recommended_movies
```

Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Some other ways to approach filtering the `unique_watched` list could be using a list comprehension or using the `filter` function with a lambda function:

```python
# list comprehension
rec_movies = [movie for movie in unique_watched if movie["host"] in user_data["subscriptions"]]

# filter with a lambda function
rec_movies = filter(lambda movie: movie["host"] in user_data["subscriptions"], unique_watched)
```

(These lines are a bit long, in practice we might split a statement like these across lines, use some extra variables, or shorten some naming to keep under the PEP8 guide of 79 characters max per line.)

More info on Lambda functions in Python in case it’s useful: https://realpython.com/python-lambda/

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,122 @@
...
+    return recommended_movies
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    
+    filtered_recommendations = []
+    for movie in  get_available_recs(user_data):
+        if movie["genre"] == get_most_watched_genre(user_data) and get_most_watched_genre(user_data):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: The data in `user_data` is not changing over the life of this function. We could reduce how much work the loop is doing by calling `get_most_watched_genre(user_data)` once before the loop and storing the value in a variable. It would also have the benefit of reducing the length of this line which is longer than 79 characters.

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,122 @@
...
+            if movie["host"] in user_data["subscriptions"]:
+                recommended_movies.append(movie)
+    return recommended_movies
 # -----------------------------------------
 # ------------- WAVE 5 --------------------
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    
+    filtered_recommendations = []
+    for movie in  get_available_recs(user_data):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Extra space after `in` operator
```suggestion
    for movie in get_available_recs(user_data):
```

File: viewing_party/party.py (hunk 18 above)
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: I would consider reversing the order of these comparisons. If `get_most_watched_genre` returns `None`, there's no reason to compare it to `movie["genre"]`. We also shouldn't use the `==` operator if we are comparing something to `None`, which will happen right now if `get_most_watched_genre` returns `None`.

If we check if `get_most_watched_genre(user_data)` exists first, then with python short circuiting if it returns `None`, it will never bother to check movie["genre"] == get_most_watched_genre(user_data).

---------------------------------------------------------

**More about testing for `None` if interested:**
In Python, when we test if a variable is pointing to `None` the PEP8 style guide recommends using `is` or `is not` for comparison rather than operators like `==`. 

`None` is a special object in Python, it’s a `singleton`, meaning there is only ever one instance of `None` in memory. All variables holding `None` are actually pointing to the same instance. By using `is` to compare against `None`, we avoid issues that could be caused by a class’s custom implementation of the comparison operators, and we gain some speed over a check using `==` since we are checking against identity rather than looking up if the variable's held type has a comparison function that should be used.
```python
if movie_title is None or genre is None or rating is None :
    return None
```
If you’d like more info on `is` vs `==` for `None`, here are a couple resources to get started =]
https://peps.python.org/pep-0008/#programming-recommendations
http://jaredgrubb.blogspot.com/2009/04/python-is-none-vs-none.html

File: viewing_party/party.py [hunk 20]
```diff
@@ -1,23 +1,122 @@
...
+            filtered_recommendations.append(movie)
+    return filtered_recommendations
+
+def get_rec_from_favorites(user_data):
 
+    recommended_movies = []
+    if "favorites" in user_data:
+        for movie in get_unique_watched(user_data):
+            if movie in user_data["favorites"]:
+                recommended_movies.append(movie)
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: What could it look like to change the if-statement to a guard clause and un-indent the main path of the function?

File: viewing_party/party.py [hunk 21]
```diff
@@ -1,23 +1,122 @@
...
+    for movie in  get_available_recs(user_data):
+        if movie["genre"] == get_most_watched_genre(user_data) and get_most_watched_genre(user_data):
+            filtered_recommendations.append(movie)
+    return filtered_recommendations
+
+def get_rec_from_favorites(user_data):
 
+    recommended_movies = []
+    if "favorites" in user_data:
+        for movie in get_unique_watched(user_data):
```
Comment (kelsey-steven-ada) [inline viewing_party/party.py]: Great code reuse in waves 4 & 5!

**Review by kelsey-steven-ada — COMMENTED** [review]
Nice work on your first pair project Carla & Weixi!

File: tests/test_wave_01.py [hunk 22]
```diff
@@ -183,12 +183,11 @@ def test_moves_movie_from_watchlist_to_watched():
     assert len(updated_data["watchlist"]) == 1
     assert len(updated_data["watched"]) == 2
     
-    raise Exception("Test needs to be completed.")
-    # *******************************************************************************************
-    # ****** Add assertions here to test that the correct movie was added to "watched" **********
-    # *******************************************************************************************
+    assert movie_to_watch in updated_data["watched"]
```
Comment (kelsey-steven-ada) [inline tests/test_wave_01.py:186]: Great use of `in` to check if `movie_to_watch` is where we expect.

We know this movie is where it belongs, but the assertion doesn't guarantee that the other movies are where we expect them. What other assertions could we add to ensure each movie is in the list we expect?

File: tests/test_wave_03.py [hunk 23]
```diff
@@ -55,12 +55,13 @@ def test_friends_unique_movies_not_duplicated():
...
     assert len(friends_unique_movies) == 3
 
-    raise Exception("Test needs to be completed.")
-    # *************************************************************************************************
-    # ****** Add assertions here to test that the correct movies are in friends_unique_movies **********
-    # **************************************************************************************************
+    assert len(friends_unique_movies) == 3
+    assert INTRIGUE_3 in friends_unique_movies  
+    assert HORROR_1 in friends_unique_movies  
+    assert FANTASY_4 in friends_unique_movies
```
Comment (kelsey-steven-ada) [inline tests/test_wave_03.py:61]: 🙌🏻
